Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialTypeSourceData.cpp

```diff
@@ -252,54 +529,183 @@ namespace AZ
+                    });
+
+                if (propertyGroupIter != propertyGroup->GetPropertyGroups().end())
+                {
+                    AZ_Error("Material source data", false, "Material property '%s' collides with a PropertyGroup with the same ID.", propertyId.GetCStr());
+                    return false;
+                }
+
+                materialTypeAssetCreator.BeginMaterialProperty(propertyId, property->m_dataType);
+                
+                if (property->m_dataType == MaterialPropertyDataType::Enum)
+                {
+                    materialTypeAssetCreator.SetMaterialPropertyEnumNames(property->m_enumValues);
+                }
+
+                for (auto& output : property->m_outputConnections)
                 {
-                    const AZStd::string& propertyName = propertyDefinition.m_name;
-                    if (!callback(groupName, propertyName, propertyDefinition))
+                    switch (output.m_type)
                     {
-                        return;
+                    case MaterialPropertyOutputType::ShaderInput:
+                    {
+                        materialTypeAssetCreator.ConnectMaterialPropertyToShaderInput(Name{output.m_fieldName});
+                        break;
+                    }
+                    case MaterialPropertyOutputType::ShaderOption:
+                    {
+                        if (output.m_shaderIndex >= 0)
+                        {
+                            materialTypeAssetCreator.ConnectMaterialPropertyToShaderOption(Name{output.m_fieldName}, output.m_shaderIndex);
+                        }
+                        else
+                        {
+                            materialTypeAssetCreator.ConnectMaterialPropertyToShaderOptions(Name{output.m_fieldName});
+                        }
+                        break;
+                    }
+                    case MaterialPropertyOutputType::Invalid:
+                        // Don't add any output mappings, this is the case when material functors are expected to process the property
+                        break;
+                    default:
+                        AZ_Assert(false, "Unsupported MaterialPropertyOutputType");
+                        return false;
                     }
                 }
-            }
-        }
 
-        void MaterialTypeSourceData::EnumeratePropertiesInDisplayOrder(const EnumeratePropertiesCallback& callback) const
-        {
-            if (!callback)
+                materialTypeAssetCreator.EndMaterialProperty();
+
+                // Parse and set the property's value...
+                if (!property->m_value.IsValid())
+                {
+                    AZ_Warning("Material source data", false, "Source data for material property value is invalid.");
+                }
+                else
+                {
+                    switch (property->m_dataType)
```

Comment:
To remove another indentation level and make this large loop a bit easier to read, I'd recommend doing a `continue;` instead

````
                // Parse and set the property's value...
                if (!property->m_value.IsValid())
                {
                    AZ_Warning("Material source data", false, "Source data for material property value is invalid.");
                    continue;
                }

                switch (property->m_dataType)
````